src/envs/playroom.py

```diff
@@ -0,0 +1,610 @@
+            init_state = self._sample_state_from_piles(piles, rng)
+            atoms = utils.abstract(init_state, self.predicates)
+            while True:  # repeat until goal is not satisfied
+                # To change later
+                goal = self._sample_goal_from_piles(num_blocks, piles, rng)
+                if not goal.issubset(atoms):
+                    break
+            tasks.append(Task(init_state, goal))
+        return tasks
+
+    def _sample_state_from_piles(self, piles: List[List[Object]],
+                                 rng: np.random.Generator) -> State:
+        data: Dict[Object, Array] = {}
+        # Create objects
+        block_to_pile_idx = {}
+        for i, pile in enumerate(piles):
+            for j, block in enumerate(pile):
+                assert block not in block_to_pile_idx
+                block_to_pile_idx[block] = (i, j)
+        # Sample pile (x, y)s
+        pile_to_xy: Dict[int, Tuple[float, float]] = {}
+        for i in range(len(piles)):
+            pile_to_xy[i] = self._sample_initial_pile_xy(
+                rng, set(pile_to_xy.values()))
+        # Create block states
+        for block, pile_idx in block_to_pile_idx.items():
+            pile_i, pile_j = pile_idx
+            x, y = pile_to_xy[pile_i]
+            z = self.table_height + CFG.playroom_block_size * (0.5 + pile_j)
+            max_j = max(j for i, j in block_to_pile_idx.values() if i == pile_i)
+            # [pose_x, pose_y, pose_z, held, clear]
+            data[block] = np.array([x, y, z, 0.0, int(pile_j == max_j)*1.0])
+        # [pose_x, pose_y, rotation, fingers], fingers start off open
+        data[self._robot] = np.array([5.0, 5.0, 0.0, 1.0])
+        # [pose_x, pose_y, open], all doors start off closed
+        data[self._door1] = np.array([30.0, 15.0, 0.0])
+        data[self._door2] = np.array([50.0, 15.0, 0.0])
+        data[self._door3] = np.array([60.0, 15.0, 0.0])
+        data[self._door4] = np.array([80.0, 15.0, 0.0])
+        data[self._door5] = np.array([100.0, 15.0, 0.0])
+        data[self._door6] = np.array([110.0, 15.0, 0.0])
+        # [pose_x, pose_y, level], light starts off off
+        data[self._dial] = np.array([125.0, 15.0, 0.0])
+        return State(data)
+
+    def _sample_initial_pile_xy(self, rng: np.random.Generator,
+                                existing_xys: Set[Tuple[float, float]]
+                                ) -> Tuple[float, float]:
+        while True:
+            x = rng.uniform(self.table_x_lb, self.table_x_ub)
+            y = rng.uniform(self.table_y_lb, self.table_y_ub)
+            if self._table_xy_is_clear(x, y, existing_xys):
+                return (x, y)
+
+    def _On_holds(self, state: State, objects: Sequence[Object]) -> bool:
+        block1, block2 = objects
+        if state.get(block1, "held") >= self.held_tol or \
+           state.get(block2, "held") >= self.held_tol:
+            return False
```

Comment:
this is the only line missing from test coverage, but i'm not sure how to deal with it??